File: src/CmdRebalance.cpp
// CmdRebalance.cpp - This file is part of QPlot

/* QPlot - Publication quality 2D graphs with dual coordinate systems
   Copyright (C) 2014  Daniel Wagenaar
  
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// CmdRebalance.C

#include "CmdRebalance.h"

#include <QDebug>
#include <math.h>
#include "Error.h"
#include "Range.h"
#include "WhichAxis.h"

static CBuilder<CmdRebalance> cbRebalance("rebalance");

constexpr double SCALETOLERANCE = 2e-3;
constexpr double SPACETOLERANCE = 1e-2; // pt


struct SpaceNeeds {
public:
  QRectF fullextent;
  QMap<QString, double> leftnondatause; // or top
  QMap<QString, double> rightnondatause; // or bottom
  QMap<QString, Range> datarange;
  QMap<QString, double> oldwidth;
public:
  SpaceNeeds();
  SpaceNeeds(Figure const &f, QStringList ids, WhichAxis const &wa) {
    // Get space available
    for (QString id: ids) {
      Panel const &p(f.panel(id));
      if (fullextent.isEmpty())
        fullextent = p.desiredExtent;
      else
        fullextent |= p.desiredExtent;
    }

    for (QString id: ids) {
      Panel const &p(f.panel(id));
      Axis const &axis(wa.axis(p));
      Range fullbb(wa.rectRange(p.fullbbox));
      Range desibb(wa.rectRange(p.desiredExtent));
      oldwidth[id] = desibb.range();
      Range databb(wa.axisPRange(axis));
      double prespace = databb.min() - fullbb.min();
      double postspace = fullbb.max() - databb.max();
      leftnondatause[id] = prespace;
      rightnondatause[id] = postspace;
      datarange[id] = Range(axis.min(), axis.max());
    }
  }
  double meanOldWidth() const {
    double ow = 0;
    for (double w: oldwidth)
      ow += w;
    return ow / oldwidth.size();
  }
  double maxNonData() const {
    return maxLeftNonData() + maxRightNonData();
  }
  double maxLeftNonData() const {
    double x = 0;
    for (double nd: leftnondatause)
      if (nd>x)
        x = nd;
    return x;
  }
  double maxRightNonData() const {
    double x = 0;
    for (double nd: rightnondatause)
      if (nd>x)
        x = nd;
    return x;
  }
  Range overallDataRange() const { // union of ranges
    Range dr;
    for (Range const &r: datarange)
      dr.unionize(r);
    return dr;
  }
};


bool CmdRebalance::usage() {
  return error("Usage: rebalance ID ...\n");
}

static QList<QStringList> blocks(Statement const &s) {
  QList<QStringList> result;
  QStringList current;
  for (int i=2; i<s.length(); i++) {
    if (s[i].typ==Token::CAPITAL) {
      current << s[i].str;
    } else if (s[i].typ==Token::DASH) {
      if (!current.isEmpty())
        result << current;
      current = QStringList();
    } else {
      return QList<QStringList>(); // fail if unexpected token
    }
  }
  if (!current.isEmpty())
    result << current; // we tolerate final DASH
  if (result.isEmpty())
    return QList<QStringList>(); // fail if nothing read
  for (QStringList const &lst: result)
    if (lst.size() < 2)
      return QList<QStringList>(); // fail if not at least two IDs in each list
  return result;
}
  

bool CmdRebalance::parse(Statement const &s) {
  if (s.length()<3)
    return usage(); // we could allow this null case, but that might confuse

  if (s[1].typ!=Token::BAREWORD)
    return usage();
  if (!(s[1].str=="x" || s[1].str=="y" || s[1].str=="xy"))
    return usage();
  if (blocks(s).isEmpty())
    return usage();
  return true;
}


static void rebalance(Figure &f, QList<QStringList> blocks,
                      WhichAxis const &wa) {
  qDebug() << "rebalance" << blocks;

  /* First step is to measure the "groups" within each block, i.e., the
     columns if wa=X or rows is wa=Y. There must be an equal number of
     groups in each block, or the operation does not make sense.
  */

  int B = blocks.size();
  if (B==0)
    return;

  QList<QList<QStringList>> groupsbyblock; // block# -> group# -> ids
  QList<QList<SpaceNeeds>> groupneedsbyblock;
  int N = -1; // groups per block
  for (QStringList block: blocks) {
    QList<QStringList> groups = wa.orderedGroups(f, block);
    if (N<0)
      N = groups.size();
    if (groups.size()!=N) {
      Error() << "Group mismatch";
      return;
    }
    groupsbyblock << groups;

    QList<SpaceNeeds> groupneeds;
    for (QStringList const &group: groups)
      groupneeds << SpaceNeeds(f, group, wa);
    groupneedsbyblock << groupneeds;
  }

  /* Next, the total nondata use within each "hypercolumn", i.e.,
     stack of columns across blocks. [I'll use "columns" which is
     correct for X; read "rows" for Y.] Also, total nondatause across
     all hypercolumns together. The rest is available for data.
  */
  QList<double> hypercolumnleftnondata;
  QList<double> hypercolumnrightnondata;
  double totalnondata;
  for (int n=0; n<N; n++) {
    double lnd = 0;
    double rnd = 0;
    for (QList<SpaceNeeds> const &needs: groupneedsbyblock) {
      double lnd1 = needs[n].maxLeftNonData();
      if (lnd1>lnd)
        lnd = lnd1;
      double rnd1 = needs[n].maxRightNonData();
      if (rnd1>rnd)
        rnd = rnd1;
    }
    hypercolumnleftnondata << lnd;
    hypercolumnrightnondata << rnd;
    totalnondata += lnd + rnd;
  }

  // Figure out full extent of each block
  QList<Range> fullextents;
  for (QList<SpaceNeeds> const &needs: groupneedsbyblock) {
    Range fe;
    for (SpaceNeeds const &need: needs)  // loop over columns
      fe.unionize(wa.rectRange(need.fullextent));
    fullextents << fe;
  }
  
  // Figure out the scale of the first block
  QList<double> scales; // will eventually contain scale for each block
  double totaldata = 0;
  for (SpaceNeeds const &need: groupneedsbyblock[0]) // loop over columns
    totaldata += need.overallDataRange().range();
  scales << (fullextents[0].range() - totalnondata) / totaldata;
  
  // Figure out new proposal for sharing space in first block
  bool trivial = true;
  QList<double> propwidth; // per column
  for (int n=0; n<N; n++) {
    SpaceNeeds const &need(groupneedsbyblock[0][n]);
    double desired = hypercolumnleftnondata[n] + hypercolumnrightnondata[n] 
      + need.overallDataRange().range() * scales[0];
    propwidth << desired;
    if (fabs(desired - need.meanOldWidth()) > SPACETOLERANCE)
      trivial = false;
  }

  // Now figure out scales for other blocks using same space distr
  for (int b=1; b<B; b++) {
    QList<SpaceNeeds> const &needs(groupneedsbyblock[b]);
    double scl = -1;
    for (int n=0; n<N; n++) { // loop over columns
      SpaceNeeds const &need(needs[n]);
      double dr = need.overallDataRange().range();
      double lnd = hypercolumnleftnondata[n];
      double rnd = hypercolumnrightnondata[n];

      double s = (propwidth[n] - lnd - rnd) / dr;
      if (scl<0 || s<scl)
        scl = s;
      if (fabs(propwidth[n] - need.meanOldWidth()) > SPACETOLERANCE)
        trivial = false;
    }
    scales << scl;
  }
    
  if (trivial)
      return;

  // Apply the scale to all panels
  for (int b=0; b<B; b++) {
    QList<SpaceNeeds> const &needs(groupneedsbyblock[b]);
    double x0 = fullextents[b].min();
    double scale = scales[b];
    for (int n=0; n<N; n++) {
      SpaceNeeds const &need(needs[n]);
      Range dr = need.overallDataRange();
      double x1 = x0 + propwidth[n];
      double xleft = x0 + hypercolumnleftnondata[n];
      for (QString id: groupsbyblock[b][n]) {
        Panel &panel(f.panelRef(id));
        QRectF ext = wa.rerect(panel.desiredExtent, x0, x1);
        f.overridePanelExtent(id, ext);
        Axis &axis(wa.axis(panel));
        axis.setDataRange(dr.min(), dr.max());
        bool rev = wa.point(axis.maprel(1)) < 0;
        double xright = xleft + dr.range()*scale;
        double px0 = rev ? xright : xleft;
        double px1 = rev ? xleft : xright;
        axis.setPlacement(wa.repoint(axis.minp(), px0),
                          wa.repoint(axis.maxp(), px1));
        qDebug() << id << dr.min() << dr.max() << xleft << xright << rev << ext;
      }
      x0 = x1;
    }
  }

  f.markFudged();
}

void CmdRebalance::render(Statement const &s, Figure &f, bool) {
  bool shareX = s[1].str.contains("x");
  bool shareY = s[1].str.contains("y");

  
  QList<QStringList> blks = blocks(s);
  for (QStringList blk: blks) {
    for (QString id: blk) {
      if (!f.hasPanel(id)) {
        Error() << "Unknown panel: " << id;
        return;
      }
    }
  }

  f.leavePanel();

  if (shareX) 
    rebalance(f, blks, WhichAxis::x());
  if (shareY) 
    rebalance(f, blks, WhichAxis::y());
}

File: backend/python/cmdrebalance.py
# cmdrebalance.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems
# Copyright (C) 2014  Daniel Wagenaar
# (same GPL licence as original)

from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from PyQt6.QtCore import QRectF

from .command import Command
from .token_ import Token
from .whichaxis import WhichAxis
from .range_ import Range
from .error import Error

if TYPE_CHECKING:
    from .statement import Statement
    from .figure import Figure
    from .panel import Panel
    from .axis import Axis


_SCALE_TOLERANCE = 2e-3
_SPACE_TOLERANCE = 1e-2  # pt


# ---------------------------------------------------------------------------
# SpaceNeeds — measurements for a group of panels along one axis
# ---------------------------------------------------------------------------

@dataclass
class SpaceNeeds:
    full_extent:       QRectF               = field(default_factory=QRectF)
    left_nondata_use:  dict[str, float]      = field(default_factory=dict)
    right_nondata_use: dict[str, float]      = field(default_factory=dict)
    data_range_map:    dict[str, Range]      = field(default_factory=dict)
    old_width:         dict[str, float]      = field(default_factory=dict)

    @staticmethod
    def build(f: Figure, ids: list[str],
              wa: WhichAxis) -> SpaceNeeds:
        sn = SpaceNeeds()

        # Get the union of desired extents across all panels in the group
        for id_ in ids:
            p = f.panel(id_)
            if sn.full_extent.isEmpty():
                sn.full_extent = p.desired_extent
            else:
                sn.full_extent = sn.full_extent.united(p.desired_extent)

        for id_ in ids:
            p = f.panel(id_)
            axis = wa.axis(p)
            desi_bb  = wa.rect_range(p.desired_extent)
            data_bb  = wa.axis_p_range(axis)
            sn.old_width[id_]         = desi_bb.range()
            sn.left_nondata_use[id_]  = data_bb.min() - desi_bb.min()
            sn.right_nondata_use[id_] = desi_bb.max() - data_bb.max()
            sn.data_range_map[id_]    = Range(axis.min(), axis.max())

        return sn

    def mean_old_width(self) -> float:
        if not self.old_width:
            return 0.0
        return sum(self.old_width.values()) / len(self.old_width)

    def max_left_nondata(self) -> float:
        return max(self.left_nondata_use.values(), default=0.0)

    def max_right_nondata(self) -> float:
        return max(self.right_nondata_use.values(), default=0.0)

    def max_nondata(self) -> float:
        return self.max_left_nondata() + self.max_right_nondata()

    def overall_data_range(self) -> Range:
        """Union of all panel data ranges in this group."""
        dr = Range()
        for r in self.data_range_map.values():
            dr.unionize(r)
        return dr


# ---------------------------------------------------------------------------
# Helper: parse ID blocks separated by dashes
# ---------------------------------------------------------------------------

def _blocks(s: Statement) -> list[list[str]]:
    """Parse 'ID ID ... - ID ID ... - ...' from token index 2 onward.
    Returns [] on any parse error or if any block has fewer than 2 IDs."""
    result: list[list[str]] = []
    current: list[str] = []

    for i in range(2, len(s)):
        if s[i].typ == Token.CAPITAL:
            current.append(s[i].str)
        elif s[i].typ == Token.DASH:
            if current:
                result.append(current)
            current = []
        else:
            return []  # unexpected token

    if current:
        result.append(current)

    if not result:
        return []
    if any(len(lst) < 2 for lst in result):
        return []

    return result


# ---------------------------------------------------------------------------
# Core algorithm
# ---------------------------------------------------------------------------

def _rebalance(f: Figure, blks: list[list[str]],
               wa: WhichAxis) -> None:
    B = len(blks)
    if B == 0:
        return

    # Step 1: group each block by position along wa, check equal group count
    groups_by_block: list[list[list[str]]] = []
    needs_by_block:  list[list[SpaceNeeds]] = []
    N = -1  # groups per block

    for block in blks:
        groups = wa.ordered_groups(f, block)
        if N < 0:
            N = len(groups)
        if len(groups) != N:
            Error() << "Group mismatch"
            return
        groups_by_block.append(groups)
        needs_by_block.append(
            [SpaceNeeds.build(f, group, wa) for group in groups]
        )

    # Step 2: compute max non-data use per hypercolumn (column across blocks)
    hc_left_nondata:  list[float] = []
    hc_right_nondata: list[float] = []
    total_nondata = 0.0

    for n in range(N):
        lnd = max(needs[n].max_left_nondata()  for needs in needs_by_block)
        rnd = max(needs[n].max_right_nondata() for needs in needs_by_block)
        hc_left_nondata.append(lnd)
        hc_right_nondata.append(rnd)
        total_nondata += lnd + rnd

    # Step 3: full extent of each block
    full_extents: list[Range] = []
    for needs in needs_by_block:
        fe = Range()
        for need in needs:
            fe.unionize(wa.rect_range(need.full_extent))
        full_extents.append(fe)

    # Step 4: scale for first block, proposed widths per column
    scales: list[float] = []
    total_data = sum(
        need.overall_data_range().range()
        for need in needs_by_block[0]
    )
    scales.append((full_extents[0].range() - total_nondata) / total_data)

    trivial = True
    prop_width: list[float] = []
    for n in range(N):
        need = needs_by_block[0][n]
        desired = (hc_left_nondata[n] + hc_right_nondata[n]
                   + need.overall_data_range().range() * scales[0])
        prop_width.append(desired)
        if math.fabs(desired - need.mean_old_width()) > _SPACE_TOLERANCE:
            trivial = False

    # Step 5: scales for remaining blocks using same column widths
    for b in range(1, B):
        needs = needs_by_block[b]
        scl = -1.0
        for n in range(N):
            need = needs[n]
            dr   = need.overall_data_range().range()
            lnd  = hc_left_nondata[n]
            rnd  = hc_right_nondata[n]
            s    = (prop_width[n] - lnd - rnd) / dr
            if scl < 0 or s < scl:
                scl = s
            if math.fabs(prop_width[n] - need.mean_old_width()) > _SPACE_TOLERANCE:
                trivial = False
        scales.append(scl)

    triv_extent = trivial

    # Step 6: apply scale to all panels
    for b in range(B):
        needs = needs_by_block[b]
        x0    = full_extents[b].min()
        scale = scales[b]

        for n in range(N):
            need = needs[n]
            dr   = need.overall_data_range()
            x1   = x0 + prop_width[n]
            x_left = x0 + hc_left_nondata[n]

            for id_ in groups_by_block[b][n]:
                panel = f.panel_ref(id_)

                if not triv_extent:
                    ext = wa.rerect(panel.desired_extent, x0, x1)
                    f.override_panel_extent(id_, ext)

                axis = wa.axis(panel)
                rev  = wa.point(axis.maprel(1)) < 0
                x_right = x_left + dr.range() * scale
                px0 = x_right if rev else x_left
                px1 = x_left  if rev else x_right

                if (math.fabs(wa.point(axis.map(dr.min())) - px0)
                        > _SPACE_TOLERANCE):
                    trivial = False
                if (math.fabs(wa.point(axis.map(dr.max())) - px1)
                        > _SPACE_TOLERANCE):
                    trivial = False

                if not trivial:
                    axis.set_data_range(dr.min(), dr.max())
                    axis.set_placement(
                        wa.repoint(axis.minp(), px0),
                        wa.repoint(axis.maxp(), px1),
                    )

            x0 = x1

    if not trivial:
        f.mark_fudged()


# ---------------------------------------------------------------------------
# Command class
# ---------------------------------------------------------------------------

@Command.register("rebalance")
class CmdRebalance(Command):
    """Rearrange axis placements so panels share the same data scale.

    Syntax:
        rebalance x  ID [ID ...] [- ID [ID ...]] ...
        rebalance y  ID [ID ...] [- ID [ID ...]] ...
        rebalance xy ID [ID ...] [- ID [ID ...]] ...

    IDs within each dash-separated block are treated as a group.
    Note: does not work for rotated axes.
    """

    def _usage(self) -> bool:
        return self.error("Usage: rebalance x|y|xy ID ...\n")

    def parse(self, s: Statement) -> bool:
        if len(s) < 3:
            return self._usage()
        if s[1].typ != Token.BAREWORD:
            return self._usage()
        if s[1].str not in ("x", "y", "xy"):
            return self._usage()
        if not _blocks(s):
            return self._usage()
        return True

    def render(self, s: Statement, f: Figure, dryrun: bool) -> None:
        share_x = "x" in s[1].str
        share_y = "y" in s[1].str

        blks = _blocks(s)
        for blk in blks:
            for id_ in blk:
                if not f.has_panel(id_):
                    Error() << "Unknown panel: " << id_
                    return

        f.leave_panel()

        if share_x:
            _rebalance(f, blks, WhichAxis.x())
        if share_y:
            _rebalance(f, blks, WhichAxis.y())
